tests/frontier/scenarios/test_scenarios.py

```diff
@@ -0,0 +1,258 @@
+@pytest.mark.parametrize(
+    # select program to debug (program, "scenario_name")
+    # program=None select all programs
+    # scenario_name="" select all scenarios
+    # Example: [ScenarioDebug(test_param=program_invalid, scenario_name="scenario_CALL_CALL")],
+    "debug",
+    # [ScenarioDebug(test_param=program_invalid, scenario_name="scenario_CALL_CALL")],
+    [ScenarioDebug(test_param=None, scenario_name="")],
+    ids=["debug"],
+)
+@pytest.mark.parametrize(
+    "operation, expected_result",
+    [
+        # invalid opcodes
+        program_invalid,
+        # static violation programs
+        program_sstore_sload,
+        program_tstore_tload,
+        program_logs,
+        # 00 - 20
+        program_all_frontier_opcodes,
+        # 30
+        program_address,
+        program_balance,
+        program_origin,
+        program_caller,
+        program_callvalue,
+        program_calldataload,
+        program_calldatasize,
+        program_calldatacopy,
+        program_codecopy_codesize,
+        program_gasprice,
+        program_ext_codecopy_codesize,
+        program_returndatasize,
+        program_returndatacopy,
+        program_extcodehash,
+        # 40
+        program_blockhash,  # fails in state mode in py-t8n
+        program_coinbase,
+        program_timestamp,
+        program_number,
+        program_difficulty_randao,
+        program_gaslimit,
+        program_chainid,
+        program_selfbalance,
+        program_basefee,
+        program_blobhash,
+        program_blobbasefee,
+        program_tload,
+        program_mcopy,
+        program_push0,
+        program_suicide,
+    ],
+)
+def test_scenarios(
+    blockchain_test: BlockchainTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    expected_result: ProgramResult,
+    debug: ScenarioDebug,
+    operation: Bytecode,
+    scenarios,
+):
+    """
+    Test given operation in different scenarios
+    Verify that it's return value equal to expected result on every scenario,
+    that is valid for the given fork.
+
+    Note: Don't use pytest parametrize for scenario production, because scenarios will be complex
+    Generate one test file for [each operation] * [each scenario] to save space
+    As well as operations will be complex too
+    """
+    tx_env = Environment()
+    tx_origin: Address = pre.fund_eoa()
+
+    blocks: List[Block] = []
+    post: dict = {}
+    for scenario in scenarios:
+        if debug.scenario_name and scenario.name != debug.scenario_name:
+            continue
+
+        if debug.test_param is not None:
+            debug_program = debug.test_param.values[0]
+            if hasattr(debug_program, "hex") and operation.hex() != debug_program.hex():
+                continue
+
+        post_storage = Storage()
+        result_slot = post_storage.store_next(1, hint=f"runner result {scenario.name}")
+
+        tx_max_gas = (
+            7_000_000
+            if hasattr(program_invalid.values[0], "hex")
```

Comment:
This is a side effect of using `pytest.param` to define the programs, it could be that they require its own class definition:
```
class Program:
    gas: int | None = None  # None means default
    ...
```